agimus_controller_ros/agimus_controller_ros/agimus_controller.py

```diff
@@ -0,0 +1,250 @@
+            )
+            if wait_for_input:
+                rclpy.get_logger().info(3, "Waiting until we receive a sensor message.")
+            rclpy.get_logger().info("Start controller")
+            self.rate.sleep()
+        return wait_for_input
+
+    def wait_buffer_has_twice_horizon_points(self):
+        while (
+            self.traj_buffer.get_size(self.point_attributes)
+            < 2 * self.params.ocp.horizon_size
+        ):
+            self.fill_buffer()
+
+    def get_next_trajectory_point(self):
+        raise RuntimeError("Not implemented")
+
+    def fill_buffer(self):
+        point = self.get_next_trajectory_point()
+        if point is not None:
+            self.traj_buffer.add_trajectory_point(point)
+        while point is not None:
+            point = self.get_next_trajectory_point()
+            if point is not None:
+                self.traj_buffer.add_trajectory_point(point)
+
+    def first_solve(self, x0):
+        # retrieve horizon state and acc references
+        horizon_points = self.traj_buffer.get_points(
+            self.params.ocp.horizon_size, self.point_attributes
+        )
+        x_plan = np.zeros([self.params.ocp.horizon_size, self.nx])
+        a_plan = np.zeros([self.params.ocp.horizon_size, self.nv])
+        for idx_point, point in enumerate(horizon_points):
+            x_plan[idx_point, :] = point.get_x_as_q_v()
+            a_plan[idx_point, :] = point.a
+
+        # First solve
+        self.ocp = OCPCrocoHPP(self.rmodel, self.cmodel, self.params.ocp)
+        self.ocp.set_planning_variables(x_plan, a_plan)
+        self.mpc = MPC(self.ocp, x_plan, a_plan, self.rmodel, self.cmodel)
+        us_init = self.mpc.ocp.u_plan[: self.params.ocp.horizon_size - 1]
+        self.mpc.mpc_first_step(x_plan, us_init, x0)
+        self.next_node_idx = self.params.ocp.horizon_size
+        if self.params.save_predictions_and_refs:
+            self.mpc.create_mpc_data()
+        _, u, k = self.mpc.get_mpc_output()
+        return u, k
+
+    def solve(self, x0):
+        self.target_translation_object_to_effector = None
+        self.fill_buffer()
+        if self.traj_buffer.get_size(self.point_attributes) > 0:
+            point = self.traj_buffer.get_points(1, self.point_attributes)[0]
+            self.last_point = point
+            new_x_ref = point.get_x_as_q_v()
+        else:
+            point = self.last_point
+            new_x_ref = point.get_x_as_q_v()
+        new_a_ref = point.a
+        self.in_world_M_effector = get_ee_pose_from_configuration(
+            self.rmodel,
+            self.rdata,
+            self.effector_frame_id,
+            new_x_ref[: self.rmodel.nq],
+        )
+        self.mpc.mpc_step(x0, new_x_ref, new_a_ref, self.in_world_M_effector)
+
+        if self.params.save_predictions_and_refs:
+            self.mpc.fill_predictions_and_refs_arrays()
+        _, u, k = self.mpc.get_mpc_output()
+        return u, k
```

Comment:
This should be handled by configuring classes introduced in https://github.com/agimus-project/agimus_controller/pull/97